Use requested depth-stencil renderbuffer formats

lexaknyazev opened the pull request:
#### 50d25492e4fe331a5e9f432c153bec5fc88cfa14
<pre>
Use requested depth-stencil renderbuffer formats
<a href="https://bugs.webkit.org/show_bug.cgi?id=242657">https://bugs.webkit.org/show_bug.cgi?id=242657</a>

Reviewed by Kenneth Russell.

Create a depth, stencil, or combined depth-stencil
renderbuffer depending on the context creation options.

Using only the requested renderbuffer features reduces
memory footprint because D24S8 may be emulated as D32FS8.

Removed applyStencilTest workaround when ANGLE is used.

Fixes (ANGLE-only):
* conformance/context/context-attributes-alpha-depth-stencil-antialias.html

* Source/WebCore/html/canvas/WebGL2RenderingContext.cpp:
(WebCore::WebGL2RenderingContext::framebufferTextureLayer):
(WebCore::WebGL2RenderingContext::renderbufferStorageMultisample):
* Source/WebCore/html/canvas/WebGLRenderingContextBase.cpp:
(WebCore::WebGLRenderingContextBase::disable):
(WebCore::WebGLRenderingContextBase::enable):
(WebCore::WebGLRenderingContextBase::framebufferRenderbuffer):
(WebCore::WebGLRenderingContextBase::framebufferTexture2D):
(WebCore::WebGLRenderingContextBase::renderbufferStorage):
(WebCore::WebGLRenderingContextBase::setFramebuffer):
* Source/WebCore/html/canvas/WebGLRenderingContextBase.h:
* Source/WebCore/platform/graphics/angle/GraphicsContextGLANGLE.cpp:
(WebCore::GraphicsContextGLANGLE::validateAttributes):
(WebCore::GraphicsContextGLANGLE::reshapeFBOs):
(WebCore::GraphicsContextGLANGLE::attachDepthAndStencilBufferIfNeeded):
(WebCore::GraphicsContextGLANGLE::validateDepthStencil):

Canonical link: <a href="https://commits.webkit.org/252900@main">https://commits.webkit.org/252900@main</a>
</pre>

kenrussell commented on Source/WebCore/platform/graphics/angle/GraphicsContextGLANGLE.cpp:
```diff
@@ -514,25 +514,23 @@ void GraphicsContextGLANGLE::readnPixelsImpl(GCGLint x, GCGLint y, GCGLsizei wid
+            m_internalDepthStencilFormat = GL_DEPTH24_STENCIL8_OES;
+        } else {
+            // Combined buffer not supported, prefer depth when both requested
+            m_internalDepthStencilFormat = GL_DEPTH_COMPONENT16;
```
Would GL_DEPTH_COMPONENT24_OES be a good choice if GL_OES_depth24 is supported? Applications tend to notice the depth buffer precision differences between 16-bit and 24-bit depth buffers.

lexaknyazev replied:
Done. FWIW, depth buffer is always 32F when ANGLE/Metal is used.

kenrussell commented on Source/WebCore/platform/graphics/angle/GraphicsContextGLANGLE.cpp:
```diff
@@ -514,25 +514,23 @@ void GraphicsContextGLANGLE::readnPixelsImpl(GCGLint x, GCGLint y, GCGLsizei wid
+    } else if (attrs.stencil)
+        m_internalDepthStencilFormat = GL_STENCIL_INDEX8;
+    else if (attrs.depth)
+        m_internalDepthStencilFormat = GL_DEPTH_COMPONENT16;
```
Here as well.

kenrussell reviewed (approved):
In general this looks good - simpler seems better. It's impractical to land multiple patches against the same WebKit bug, so please plan on this one superseding the patch already attached to https://bugs.webkit.org/show_bug.cgi?id=223734 . In particular, I think it's worth removing the mention of landing the earlier patch from this pull request's description, and commenting on https://bugs.webkit.org/show_bug.cgi?id=223734 indicating why you're planning on a different direction than the originally proposed patch.

kenrussell reviewed (approved):
Nice cleanup of longstanding messy code. r+

lexaknyazev committed:
Use requested depth-stencil renderbuffer formats
https://bugs.webkit.org/show_bug.cgi?id=242657

Reviewed by Kenneth Russell.

Create a depth, stencil, or combined depth-stencil
renderbuffer depending on the context creation options.

Using only the requested renderbuffer features reduces
memory footprint because D24S8 may be emulated as D32FS8.

Removed applyStencilTest workaround when ANGLE is used.

Fixes (ANGLE-only):
* conformance/context/context-attributes-alpha-depth-stencil-antialias.html

* Source/WebCore/html/canvas/WebGL2RenderingContext.cpp:
(WebCore::WebGL2RenderingContext::framebufferTextureLayer):
(WebCore::WebGL2RenderingContext::renderbufferStorageMultisample):
* Source/WebCore/html/canvas/WebGLRenderingContextBase.cpp:
(WebCore::WebGLRenderingContextBase::disable):
(WebCore::WebGLRenderingContextBase::enable):
(WebCore::WebGLRenderingContextBase::framebufferRenderbuffer):
(WebCore::WebGLRenderingContextBase::framebufferTexture2D):
(WebCore::WebGLRenderingContextBase::renderbufferStorage):
(WebCore::WebGLRenderingContextBase::setFramebuffer):
* Source/WebCore/html/canvas/WebGLRenderingContextBase.h:
* Source/WebCore/platform/graphics/angle/GraphicsContextGLANGLE.cpp:
(WebCore::GraphicsContextGLANGLE::validateAttributes):
(WebCore::GraphicsContextGLANGLE::reshapeFBOs):
(WebCore::GraphicsContextGLANGLE::attachDepthAndStencilBufferIfNeeded):
(WebCore::GraphicsContextGLANGLE::validateDepthStencil):

Canonical link: https://commits.webkit.org/252900@main
